### PR TITLE
8329258: TailCall should not use frame pointer register for jump target

### PR DESCRIPTION
**Pull request by TobiHartmann**
Applying @danielogh's patch (see description of [JDK-8329258](https://bugs.openjdk.org/browse/JDK-8329258)) to enable `StressGCM` / `StressLCM` for stub compilations triggers a crash on AArch64. The problem is that the register allocator uses `R29` (`rfp`) which is usually used for the frame pointer to hold the `TailCall` `exc_target` when generating the `_slow_arraycopy_Java` stub:
https://github.com/openjdk/jdk/blob/6dfb8120c270a76fcba5a5c3c9ad91da3282d5fa/src/hotspot/share/opto/generateOptoStub.cpp#L258-L264

With `StressGCM` / `StressLCM` the register initialization is scheduled early and `R29` is corrupted by the `MachEpilogNode` which is opaque to the register allocator and inserted right before the `TailCall`:
https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/share/opto/output.cpp#L320-L326 

```
028 mov R29, 0x0000ffff78cc0080 # ptr

[...]

098     # pop frame 16
        ldp  lr, rfp, [sp,#0]              <- Epilog kills rfp (and lr + sp)
        add  sp, sp, #16

[...]

0a0 br R29 # R12 holds method
```

As a result, we jump to a "garbage" location. See [bad code](https://bugs.openjdk.org/secure/attachment/108835/BAD_slow_arraycopy_Java.txt) vs. [good code](https://bugs.openjdk.org/secure/attachment/108834/GOOD_slow_arraycopy_Java.txt).

On x86, we use `no_rbp_RegP` instead:
https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/cpu/x86/x86_64.ad#L2564-L2566 https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/cpu/x86/x86_64.ad#L12470

I implemented the same on AArch64.

I think other platforms are affected as well but I don't have the hardware to test there. @offamitkumar (S390), @TheRealMDoerr (PPC), @RealFYang (RISC-V), @bulasevich (ARM32), could you please have a look?

I also wondered if `R29` shouldn't be a callee-save (SOE) register in the C calling convention?
https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/cpu/aarch64/aarch64.ad#L139-L140 On x86_64, `RBP` is SOE:
https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/cpu/x86/x86_64.ad#L86-L87

`TestTailCallInArrayCopyStub.java` will only work once [JDK-8330016](https://bugs.openjdk.org/browse/JDK-8330016) is integrated.

Thanks,
Tobias

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329258](https://bugs.openjdk.org/browse/JDK-8329258): TailCall should not use frame pointer register for jump target (**Bug** - P2)

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) ⚠️ Review applies to [160ea006](https://git.openjdk.org/jdk/pull/18716/files/160ea00692a41806ce590a99b84a84de39c43f92)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Contributors
 * Fei Yang `<fyang@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18716/head:pull/18716` \
`$ git checkout pull/18716`

Update a local copy of the PR: \
`$ git checkout pull/18716` \
`$ git pull https://git.openjdk.org/jdk.git pull/18716/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18716`

View PR using the GUI difftool: \
`$ git pr show -t 18716`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18716.diff">https://git.openjdk.org/jdk/pull/18716.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18716#issuecomment-2047454787)

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
I don't think PPC64 or s390 are affected. These platforms don't have a frame pointer register. Both currently use a fixed register for the jump target (the same one as for inline caches).

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:16193]: This is a somewhat opaque comment. Please say that a `MachEpilogNode` has already been emitted, so the current rfp contains a value that belongs to the caller of this method, so it must not be used.

**Comment by dean-long** [comment]
> I also wondered if R29 shouldn't be a callee-save (SOE) register in the C calling convention?

Maybe for documentation purposes, but I don't think it would have any effect on generated code.

**Comment by offamitkumar** [comment]
I have tested on s390x, test passes there as well.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/aarch64.ad]: Thanks Andrew. I adjusted the comments, please let me know what you think.

**Comment by TobiHartmann** [comment]
@TheRealMDoerr , @offamitkumar thanks for checking!

> Both currently use a fixed register for the jump target (the same one as for inline caches).

Looks to me as if no fixed register but `iRegP` is used for `jump_target`:
https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/cpu/s390/s390.ad#L9505

But the `MachEpilog` code does look safe, as you said, there is no frame pointer register.

RISC-V and ARM32 look affected to me though. Waiting for confirmation. Please note that `TestTailCallInArrayCopyStub.java` needs [JDK-8330016](https://bugs.openjdk.org/browse/JDK-8330016) to trigger the issue.

> Maybe for documentation purposes, but I don't think it would have any effect on generated code.

Thanks for checking. @dean-long, @theRealAph any preferences? Or should we leave it as is?

**Comment by dean-long** [comment]
> Thanks for checking. @dean-long, @theRealAph any preferences? Or should we leave it as is?

I would leave it.  I found this code that makes it sound like NS is better:

https://github.com/openjdk/jdk/blob/70944ca54ad0090c734bb5b3082beb33450c4877/src/hotspot/share/opto/lcm.cpp#L910

because we know how to find oops even if they are in the FP register, so there is no need to kill it across a non-leaf runtime call.

**Comment by TobiHartmann** [comment]
Okay, let's leave it then.

**Comment by RealFYang** [comment]
@TobiHartmann : Yes, you are right. This issue also triggers on linux-riscv64 if I force allocating the frame pointer (x8) for `jump_target`. And I have prepared a similar fix. Could you please add it? Thanks.
[18716-riscv.diff.txt](https://github.com/openjdk/jdk/files/14942269/18716-riscv.diff.txt)

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (RealFYang) [inline src/hotspot/cpu/aarch64/aarch64.ad:16193]: Suggestion: s/rbp/rfp/

Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/aarch64.ad:16193]: Good catch! Will fix.

Comment (RealFYang) [reply src/hotspot/cpu/aarch64/aarch64.ad:16193]: But the comment here says `it will reset rbp to the caller state`. I think `rbp` is x86-specific, right?

**Comment by TobiHartmann** [comment]
Thanks for the review and the patch, @RealFYang! I added it.

**Review by robcasloz — APPROVED** [review]
Looks good, thanks for fixing this!

**Comment by TobiHartmann** [comment]
Thanks for the review, Roberto!

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:16193]:
```suggestion
// emitted just above the TailCall which has reset rfp to the caller state.
```

Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/aarch64.ad]: Thanks. I updated the comments accordingly.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/aarch64.ad]: Right, I noticed in the meantime and updated my comment above. Thank you!

**Comment by TheRealMDoerr** [comment]
> @TheRealMDoerr , @offamitkumar thanks for checking!
> 
> > Both currently use a fixed register for the jump target (the same one as for inline caches).
> 
> Looks to me as if no fixed register but `iRegP` is used for `jump_target`:
> 
> https://github.com/openjdk/jdk/blob/b49ba426a721db5926ac1b45d573d468389d479c/src/hotspot/cpu/s390/s390.ad#L9505
> 
> But the `MachEpilog` code does look safe, as you said, there is no frame pointer register.
> 

Oh, right. I had looked at the wrong operand. But it's still safe because there is no frame pointer register on these 2 platforms.
The test has passed on PPC64 with both patches applied. Thanks!

**Comment by TobiHartmann** [comment]
Thanks again for confirming, Martin.

**Comment by bulasevich** [comment]
> I think other platforms are affected as well but I don't have the hardware to test there. 
> @bulasevich (ARM32), could you please have a look?

Hi. I checked ARM32. R11 (FP) is a common register that is not just dedicated solely to the frame pointer. And with a given test and patch I can not reproduce SIGSEGV on ARM32 platform. So I think ARM32 is not affected.

**Comment by dean-long** [comment]
> > I think other platforms are affected as well but I don't have the hardware to test there.
> > @bulasevich (ARM32), could you please have a look?
> 
> Hi. I checked ARM32. R11 (FP) is a common register that is not just dedicated solely to the frame pointer. And with a given test and patch I can not reproduce SIGSEGV on ARM32 platform. So I think ARM32 is not affected.

It looks like ARM32 is safe because the rules use specific, "bound" registers.  If they used something generic like iRegP, I think the callee-saved R11 would still be in danger of getting trashed.

**Comment by TobiHartmann** [comment]
Thanks for checking, @bulasevich and @dean-long. The current patch should be sufficient then.

Could I please get a second review? (@theRealAph?)

**Review by theRealAph — APPROVED** [review]
Thanks. Sorry for my slow reply.

**Pull request closed without merging**
